File: elixir/kindergarten-garden/garden.exs
defmodule Garden do
  @doc """
    Accepts a string representing the arrangement of cups on a windowsill and a
    list with names of students in the class. The student names list does not
    have to be in alphabetical order.

    It decodes that string into the various gardens for each student and returns
    that information in a map.
  """

  @default_students ~w(alice bob charlie david eve fred ginny harriet ileana joseph kincaid larry)a

  @spec info(String.t(), list) :: map
  def info(info_string, student_names \\ @default_students) do
    do_info(info_string, Enum.sort(student_names))
  end

  defp do_info(info_string, students) do
    info_string
    |> build_garden(students)
    |> populate(default_garden(students))
  end

  defp default_garden(students) do
    Enum.into(students, %{}, fn name -> {name, {}} end)
  end

  defp build_garden(info_string, students) do
    String.split(info_string, "\n")
    |> Enum.flat_map(&(parse_row(&1, students)))
    |> Enum.group_by(&(elem(&1, 1)), &(elem(&1, 0)))
  end

  defp parse_row(row, students) do
    row
    |> String.graphemes
    |> Enum.map(&parse_plant/1)
    |> Enum.chunk(2)
    |> Enum.zip(students)
  end

  defp parse_plant("G"), do: :grass
  defp parse_plant("C"), do: :clover
  defp parse_plant("R"), do: :radishes
  defp parse_plant("V"), do: :violets

  defp populate(garden, default) do
    Enum.reduce(garden, default, fn {student, plants}, garden ->
      Map.put(garden, student, plants |> List.flatten |> List.to_tuple)
    end)
  end
end

File: scala/kindergarten-garden/src/main/scala/garden.scala
class Garden(students: List[String], plots: String) {

  val plants: Map[String, List[Plant]] = {
    Garden.build(students.sorted.toArray, plots.split("\n"))
  }

  def getPlants(name: String): List[Plant] = {
    plants(name)
  }

}

object Garden {

  def apply(students: List[String], plots: String): Garden = new Garden(students, plots)

  def defaultGarden(plots: String): Garden = Garden(DefaultStudents, plots)

  def build(students: Array[String], plots: Seq[String]): Map[String, List[Plant]] = {
    var garden = Map.empty[String, List[Plant]].withDefaultValue(List.empty[Plant])

    for (plot <- plots; (plant, i) <- plot.zipWithIndex) {
      val student = students(i/2)
      garden += (student -> (garden(student) :+ PlantNames(plant)))
    }

    garden
  }

  val DefaultStudents = List("Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny",
    "Harriet", "Ileana", "Joseph", "Kincaid", "Larry")

  val PlantNames: Map[Char, Plant] = Map(('C' -> Plant.Clover), ('G' -> Plant.Grass),
    ('R' -> Plant.Radishes), ('V' -> Plant.Violets))
}

sealed trait Plant
object Plant {
  case object Clover extends Plant
  case object Grass extends Plant
  case object Radishes extends Plant
  case object Violets extends Plant
}
